drivers/input/ajoystick.c

```diff
@@ -722,6 +720,19 @@ static int ajoy_poll(FAR struct file *filep, FAR struct pollfd *fds,
 
               opriv->ao_fds[i] = fds;
               fds->priv = &opriv->ao_fds[i];
+
+              /* Report if the event is pending */
+
+              if (opriv->ao_pollpending)
+                {
+                  fds->revents |= (fds->events & POLLIN);
+                  if (fds->revents != 0)
+                    {
+                      iinfo("Report events: %02x\n", fds->revents);
+                      nxsem_post(fds->sem);
```

Comment:
Why do we need to `nxsem_post(fds->sem);` here?

Reply:
Notify the caller there is event pending for read to avoid the caller block in poll/select. This is a normal action in any xxx_poll(e.g. https://github.com/apache/incubator-nuttx/blob/master/drivers/input/touchscreen_upper.c#L334-L342).

Reply:
Ok. But in compare to `touchscreen_upper.c` case we still consume `opriv->ao_fds[i] = fds;`. I think this should be fixed. `opriv->ao_fds[i] = fds;` should be consumed only `if (fds->revents == 0)`.

Reply:
No, the protocol require that fds must save into ao_fds if it return true regardless the semaphore post or not. You can see touchscreen_upper.c do the same thing:
https://github.com/apache/incubator-nuttx/blob/master/drivers/input/touchscreen_upper.c#L319-L324
poll will call ajoy_poll with setup == false immediately after return.

Reply:
> No, the protocol require that fds must save into ao_fds if it return true regardless the semaphore post or not. You can see touchscreen_upper.c do the same thing:
> https://github.com/apache/incubator-nuttx/blob/master/drivers/input/touchscreen_upper.c#L319-L324
> poll will call ajoy_poll with setup == false immediately after return.

Thank you, I missed that.
But what I see in the touchscreen example is that fd sem count is checked in prior to sem post and all poll waiters are notified. But in current changes it seems like behavior is different, isn't it?

Reply:
Since the semaphore used by poll always initialize before xxx_poll, so there is no real difference here. But for the semaphore used to wake up xxx_read, post without check may wake up read thread unexpectedly.

Reply:
What about notification logic? This is still not answered. In `touch_poll` we have:
```
      for (i = 0; i < CONFIG_INPUT_TOUCHSCREEN_NPOLLWAITERS; i++)
        {
          if (NULL == upper->fds[i])
            {
              upper->fds[i] = fds;
              fds->priv = &upper->fds[i];
              break;
            }
        }
...
      if (eventset)
        {
          touch_notify(upper, eventset);
        }
```
and `touch_notify` iterates `for (i = 0; i < CONFIG_INPUT_TOUCHSCREEN_NPOLLWAITERS; i++)` again,
With this logic we just notify until the first free entry met.

Reply:
Got it! LGTM!